Deprecate ECS settings applier

arnaldo2792 opened the pull request:
**Issue number:**

Closes #3621

**Description of changes:**

This deprecates the ECS settings applier and migrates all the ECS configurations to use systemd drop-ins. 


The `ecs.config` template file was removed, and replaced with systemd drop-ins for the `ecs.service`. This will allow downstream builders consume the configurations and build upon them if needed. The new configuration files include:

- `00-defaults.conf`: immutable configuration file with the default hard-coded configurations used for Bottlerocket
- `10-base.conf`: mutable configuration file generated by the API server
- `20-nvidia.conf`: immutable configuration file with the hard-coded configuration to enable NVIDIA support

With this PR, the `ecs-agent-base` and `ecs-agent-nvidia` packages were created to pull down the correct configuration while building the variant. Variants were updated accordingly to use the respective file.

In the new `ecs-base-conf` template file, most of the `{{#if}}` statements were removed in favor of using helpers like `default`. There are two ECS boolean configurations in the API system whose values are inverted:

- `ECS_PRIVILEGED_DISABLED` -> `allow-privileged-containers`
- `ECS_DISABLE_IMAGE_CLEANUP` -> `image-cleanup-enabled`

I tried to use the native `not` helper to generate the correct values for the settings, e.g. `{{not (default <setting>)}}`. However, the `default` helper will transform the `boolean` types to `string` when passed to the `not` helper. This results in the `not` helper always returning `true`, since not-empty strings are considered "truthy".  I added the `negate_or_else` handlebars helper, which will render the passed configuration or a default value.

**Testing done:**

For aws-ecs-2:

- [X] Run internal suite of tests
- [X] Test ECS_TASK_METADATA_RPS_LIMIT
- [X] Test that both ECS_PRIVILEGED_DISABLED and ECS_DISABLE_IMAGE_CLEANUP are set as expected
- [X] Test that the value for `ECS_AVAILABLE_LOGGING_DRIVERS` is a valid JSON string, and that the ECS agent parses it
- [x] Test `ECS_ENABLE_CONTAINER_METADATA`

```bash
bash-5.1# docker exec 7b059bad5345 sh -c '[ -f ${ECS_CONTAINER_METADATA_FILE} ] && echo Exists!'                                                                                                                                                                                          
Exists!
```

- [x] Migration testing
    - [x] Verified all the configuration files exist on upgrade
    - [x] Verified both `ecs.config` and `ecs.config.json` exist on downgrade 

For aws-ecs-1: 

- [x] Run internal suite of tests

All the other tests are omitted since the configuration file and API definition are the same for both variants.

For aws-ecs-2-nvidia:

- [X] Run smoke test

For aws-ecs-1-nvidia:

- [x] Run smoke test

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
ecs-agent: migrate all ECS settings to drop-ins

With this, all the API settings to configure the ECS agent are now
created as systemd drop-in files. Two subpackages were created to
modularize the configuration.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 committed:
os: remove ecs-settings-applier

Now that all the configurations are included in systemd drop-ins, this
binary isn't needed

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
(Forced push added unit tests)

arnaldo2792 committed:
migrations: add migrations needed to remove ECS settings applier

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
(forced push adds the migration and fixes the implementation of one of the schnauzer helpers)

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -336,8 +359,19 @@ install -p -m 0644 %{S:300} %{buildroot}%{_cross_datadir}/logdog.d
+posix.symlink('ecs.generic.config', "%{_cross_templatedir}/ecs.config")
+
+%post nvidia -p <lua>
+posix.symlink('ecs.nvidia.config', "%{_cross_templatedir}/ecs.config")
```
nit: scriptlets like `%post` go above `%files`.

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -262,7 +278,14 @@ install -D -p -m 0644 %{S:101} %{S:200} %{buildroot}%{_cross_unitdir}
+sed -e 's|__ENABLE_GPU_SUPPORT__|false|g' %{S:104} > ecs.generic.config
+
+install -d %{buildroot}%{_cross_templatedir}
+install -p -m 0644 ecs.nvidia.config ecs.generic.config %{buildroot}%{_cross_templatedir}
```
On the one hand, I appreciate that we're using the same config template, just with a slight variation.

On the other - we're not going to get to a truly extensible and composable set of packages unless we find ways for most packages like this to install side-by-side and build up the base config. Otherwise everyone will have to copy/paste entire templates just to make minor changes.

The most systemd-native approach would be:
* have a series of templates that render `Environment=` lines to drop-ins
   * `/etc/systemd/system/ecs.service.d/00-base.conf`
   * `/etc/systemd/system/ecs.service.d/10-gpu.conf`
* have the `ecs-config` template in one package
* have the `ecs-config-nvidia` a different package that depends on the base one

I don't think we're quite ready to do this today - we don't have different models for `aws-ecs-2` and `aws-ecs-2-nvidia`, and it would be onerous to introduce them here. It's also a bit unfair to single out ECS for this treatment when the same problem exists for kubelet and containerd. But ECS is actually pretty close to the ideal behavior where we can just stack drop-ins to achieve the desired goal of extensible configuration.

The drop-in architecture would also let us distinguish between hard-coded values (which could be in a fixed drop-in or else just an EnvironmentFile) and templated, settings-derived values.

bcressey commented on packages/ecs-agent/ecs.config.in:
```diff
@@ -0,0 +1,66 @@
+
+{{! Boolean configurations whose values are inverted in the API }}
+ECS_PRIVILEGED_DISABLED="{{not_ng (default false settings.ecs.allow-privileged-containers)}}"
+ECS_DISABLE_IMAGE_CLEANUP="{{not_ng (default true settings.ecs.image-cleanup-enabled)}}"
```
Maybe an `negate_or_else <default_bool_value> <setting>`  helper? `not_ng` has a pretty elaborate contract for something that we really only want to operate on booleans, which in part seems to come from chaining into together with default. If we pass the desired default as a parameter, we can also make the helper throw an error if it's given a non-boolean.

bcressey commented on packages/ecs-agent/ecs.config.in:
```diff
@@ -0,0 +1,66 @@
+ECS_DISABLE_IMAGE_CLEANUP="{{not_ng (default true settings.ecs.image-cleanup-enabled)}}"
+
+{{! Check if the setting is present, otherwise "null" is passed to the ECS agent which will cause
+a failure }}
```
nit: I don't think comments like these are that useful

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -14,7 +14,7 @@ use serde_json::value::Value;
 use snafu::{OptionExt, ResultExt};
 use std::borrow::Borrow;
-use std::collections::HashMap;
+use std::collections::{BTreeMap, HashMap};
```
I'd recommend splitting this commit into one helper per commit. There's kind of a lot going on.

bcressey commented on packages/ecs-agent/ecs.config.in:
```diff
@@ -0,0 +1,66 @@
+a failure }}
+{{#if settings.ecs.logging-drivers }}
+ECS_AVAILABLE_LOGGING_DRIVERS='{{json_string settings.ecs.logging-drivers}}'
+{{/if}}
```
Do we really need a helper for this? Can't we iterate over the vec and write the comma-separated output, the way we do for `ECS_ENGINE_AUTH_DATA`?

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1328,6 +1339,216 @@ pub fn etc_hosts_entries(
+        (Value::Null, Value::Number(burst)) => {
+            format!("{},{}", DEFAULT_ECS_METADATA_SERVICE_RPS, burst)
+        }
+        // It is unlikely that this is reached with values of type different than Null, since the
+        // API validates the types
+        (_, _) => format!(
+            "{},{}",
+            DEFAULT_ECS_METADATA_SERVICE_RPS, DEFAULT_ECS_METADATA_SERVICE_BURST
+        ),
```
This comment is only true if the template is well-formed and only number-valued settings are passed to the helper. IMO it should be a rendering error instead for unknown types.

But you do need the `(Value::Null, Value::Null)` case.

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1328,6 +1339,216 @@ pub fn etc_hosts_entries(
+        ),
+    };
+
+    // Write the negated value
+    out.write(&output).context(error::TemplateWriteSnafu {
+        template: template_name.to_owned(),
+    })?;
```
Comment is incorrect - it's not a negated value

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1328,6 +1339,216 @@ pub fn etc_hosts_entries(
+    Ok(())
+}
+
+/// This helper writes the passed value as a valid JSON string
+///
+/// In handlebars, the values passed to "{{}}" are written as plain strings, instead of their JSON
+/// representation. This helper takes the given value and serializes it as JSON. For example, for
+/// ["one", "two"] the returned value will be ["one", "two"] which otherwise will be written as
+/// [one, two] with plain handlebars.
+pub fn json_string(
```
This feels like too much to take on in a helper.

At an extreme, we could have a JSON config file that consisted of nothing but:
```
{{json_string settings.my-daemon}}
```

I don't like that because as a rule, I don't think Bottlerocket's API should be shaped exactly like any particular agent's config file. (Among other reasons: if the config file changes its structure, we don't want to incur an API break.)

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1328,6 +1339,216 @@ pub fn etc_hosts_entries(
+
+/// This helper returns the full map of ECS instance attributes
+///
+/// This helper takes the ECS instance attributes provided by the API, and appends an extra
+/// instance attribute that includes the variant id. The helper writes the JSON representation of
+/// the final map of instance attributes.
```
I'm really skeptical about this one. Can we build up the list on the template side and then append `os.variant_id` at the end?

arnaldo2792 committed:
schnauzer: add negate_or_else template helper

This adds the new `negate_or_else` helper which takes a boolean value
and applies a boolean not to it.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 committed:
schnauzer: add ecs_metadata_service_limits template helper

This adds the `ecs_metadata_service_limits` template helper used to render
properly formatted values for the `ECS_METADATA_SERVICE_LIMITS`
configuration for the ECS agent.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push includes:
- Use drop-ins instead of configuration files
- Remove unnecessary handlebars helpers
- Split commits that add handlebars helpers

arnaldo2792 commented:
(forced push includes rebase to resolve conflicts)

arnaldo2792 commented:
Forced push to address one comment about the `ecs_metadata_service_limits` helper.

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -86,6 +88,21 @@ Requires: %{_cross_os}amazon-ssm-agent
+%{summary}.
+
+%package nvidia
+Summary: Files specific for the NVIDIA support
```
nit: keep summary wording consistent
```suggestion
Summary: NVIDIA specific configuration files for the ECS agent
```

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -86,6 +88,21 @@ Requires: %{_cross_os}amazon-ssm-agent
 %description
 %{summary}.
 
+%package base
```
nit: since `ecs-agent-config` is a more "obvious" name than `ecs-agent-base`:
```suggestion
%package config
```

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -86,6 +88,21 @@ Requires: %{_cross_os}amazon-ssm-agent
+%description base
+%{summary}.
+
+%package nvidia
```
if you change the name of `base`:
```suggestion
%package nvidia-config
```

bcressey commented on packages/ecs-agent/ecs-tmpfiles.conf:
```diff
@@ -3,3 +3,4 @@ d /var/log/ecs 0755 root root
 d /var/log/ecs/exec 0755 root root -
 R /var/ecs/managed-agents - - - -
 d /var/ecs/managed-agents/execute-command/config 0750 root root -
+d /etc/systemd/system/ecs.service.d 0700 root root -
```
Does this work as expected? Usually `tmpfiles` runs after `local-fs.target` while the mount unit would run before. `systemd` might be creating the mount path for us without this.

bcressey commented on packages/systemd/systemd-tmpfiles.conf:
```diff
@@ -6,3 +6,4 @@ z /var/lib/systemd/random-seed 600 root root -
 R /var/lib/systemd/linger
 D /var/lib/systemd/linger 0700 root root -
 d /etc/sysctl.d 0700 root root -
+d /etc/systemd/system/ 0700 root root -
```
Isn't this created automatically?

bcressey commented on packages/ecs-agent/ecs-base-conf:
```diff
@@ -0,0 +1,59 @@
+Environment=ECS_LOGLEVEL="{{settings.ecs.loglevel}}"
+Environment=ECS_NUM_IMAGES_DELETE_PER_CYCLE="{{default 5 settings.ecs.image-cleanup-delete-per-cycle}}"
+Environment=ECS_RESERVED_MEMORY="{{default 0 settings.ecs.reserved-memory}}"
+Environment=ECS_TASK_METADATA_RPS_LIMIT="{{ecs_metadata_service_limits settings.ecs.metadata-service-rps settings.ecs.metadata-service-burst}}"
```
For consistency with the rest of the template, where the defaults are visible, you could pass the defaults here as well, e.g.:
```suggestion
Environment=ECS_TASK_METADATA_RPS_LIMIT="{{ecs_metadata_service_limits 40 60 settings.ecs.metadata-service-rps settings.ecs.metadata-service-burst}}"
```

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1368,6 +1377,129 @@ pub fn etc_hosts_entries(
+    let default_value = get_param(helper, 0)?;
+    let value_to_negate = get_param(helper, 1)?;
+
+    let default = match default_value {
```
GitHub is highlighting `default` as keyword - maybe pick a different identifier.

arnaldo2792 commented:
Forced push includes:

- Fix description in subpackage for NVIDIA
- Remove unnecessary lines in tmpfiles.d
- Renamed `base` package to `config`
- Renamed `default` variable in helper to `fallback`

arnaldo2792 commented:
(Forced push includes a missing fix to rename `nvidia` to `nvidia-config`)

arnaldo2792 commented:
(Forced push to rebase onto develop)
